bolt/include/bolt/Passes/NonPacProtectedRetAnalysis.h

```diff
@@ -0,0 +1,266 @@
+    if (BF != RHS.BF)
+      return BF < RHS.BF;
+    return Offset < RHS.Offset;
+  }
+  operator MCInst &() const {
+    assert(BF != nullptr);
+    return *(BF->getInstructionAtOffset(Offset));
+  }
+
+  uint64_t getOffset() const { return Offset; }
+
+  uint64_t getAddress() const { return BF->getAddress() + getOffset(); }
+};
+
+raw_ostream &operator<<(raw_ostream &OS, const MCInstInBFReference &);
+
+struct MCInstReference {
+  enum Kind { FunctionParent, BasicBlockParent };
+  Kind ParentKind;
+  union U {
+    MCInstInBBReference BBRef;
+    MCInstInBFReference BFRef;
+    U(MCInstInBBReference BBRef) : BBRef(BBRef) {}
+    U(MCInstInBFReference BFRef) : BFRef(BFRef) {}
+  } U;
+  MCInstReference(MCInstInBBReference BBRef)
+      : ParentKind(BasicBlockParent), U(BBRef) {}
+  MCInstReference(MCInstInBFReference BFRef)
+      : ParentKind(FunctionParent), U(BFRef) {}
+  MCInstReference(class BinaryBasicBlock *BB, int64_t BBIndex)
+      : MCInstReference(MCInstInBBReference(BB, BBIndex)) {}
+  MCInstReference(class BinaryFunction *BF, uint32_t Offset)
+      : MCInstReference(MCInstInBFReference(BF, Offset)) {}
+
+  bool operator<(const MCInstReference &RHS) const {
+    if (ParentKind != RHS.ParentKind)
+      return ParentKind < RHS.ParentKind;
+    switch (ParentKind) {
+    case BasicBlockParent:
+      return U.BBRef < RHS.U.BBRef;
+    case FunctionParent:
+      return U.BFRef < RHS.U.BFRef;
+    }
+    llvm_unreachable("");
+  }
+
+  bool operator==(const MCInstReference &RHS) const {
+    if (ParentKind != RHS.ParentKind)
+      return false;
+    switch (ParentKind) {
+    case BasicBlockParent:
+      return U.BBRef == RHS.U.BBRef;
+    case FunctionParent:
+      return U.BFRef == RHS.U.BFRef;
+    }
+    llvm_unreachable("");
+  }
+
+  operator MCInst &() const {
+    switch (ParentKind) {
+    case BasicBlockParent:
+      return U.BBRef;
+    case FunctionParent:
+      return U.BFRef;
+    }
+    llvm_unreachable("");
+  }
+
+  uint64_t getAddress() const {
+    switch (ParentKind) {
+    case BasicBlockParent:
+      return U.BBRef.getAddress();
+    case FunctionParent:
+      return U.BFRef.getAddress();
+    }
+    llvm_unreachable("");
+  }
+
+  BinaryFunction *getFunction() const {
+    switch (ParentKind) {
+    case FunctionParent:
+      return U.BFRef.BF;
+    case BasicBlockParent:
+      return U.BBRef.BB->getFunction();
+    }
+    llvm_unreachable("");
+  }
+
+  BinaryBasicBlock *getBasicBlock() const {
+    switch (ParentKind) {
+    case FunctionParent:
+      return nullptr;
+    case BasicBlockParent:
+      return U.BBRef.BB;
+    }
+    llvm_unreachable("");
+  }
+};
+
+raw_ostream &operator<<(raw_ostream &OS, const MCInstReference &);
+
+struct GeneralDiagnostic {
+  std::string Text;
+  GeneralDiagnostic(const std::string &Text) : Text(Text) {}
+  bool operator==(const GeneralDiagnostic &RHS) const {
+    return Text == RHS.Text;
+  }
+};
+
+struct NonPacProtectedRetAnnotation {
```

Comment:
With the changes in commit f44f9bf, the code now has quite a few different classes (and variables?) etc with long names containing the string "NonPacProtectedRet".
Could this be improved by introduced a namespace `NonPacProtectedRet` or `NonPacProtectedRetAnalysis`, so that we can remove that part from class names for classes in the namespace. This should make the code substantially easier to read?

Reply:
Now done in commit  8d4ac8e